web/src/compose_ui.js

```diff
@@ -646,6 +605,61 @@ export function format_text($textarea, type, inserted_content) {
     }
 }
 
+// Check if the selection is already surrounded by syntax
+function is_selection_formatted(syntax_start, syntax_end, range, text) {
+    return (
+        range.start >= syntax_start.length &&
+        text.length - range.end >= syntax_end.length &&
+        text.slice(range.start - syntax_start.length, range.start) === syntax_start &&
+        text.slice(range.end, range.end + syntax_end.length) === syntax_end
+    );
+}
+
+// Check if selected text itself has syntax inside it.
+function is_inner_text_formatted(syntax_start, syntax_end, range, selected_text) {
+    return (
+        range.length >= syntax_start.length + syntax_end.length &&
+        selected_text.slice(0, syntax_start.length) === syntax_start &&
+        selected_text.slice(-syntax_end.length) === syntax_end
+    );
+}
+
+function add_formatting(syntax_start, syntax_end, field, range, text, selected_text) {
+    let linebreak_start = "";
+    let linebreak_end = "";
+    if (syntax_start[0] === "\n") {
+        linebreak_start = "\n";
+    }
+    if (syntax_end.at(-1) === "\n") {
+        linebreak_end = "\n";
+    }
+    if (is_selection_formatted(syntax_start, syntax_end, range, text)) {
+        text =
+            text.slice(0, range.start - syntax_start.length) +
+            linebreak_start +
+            text.slice(range.start, range.end) +
+            linebreak_end +
+            text.slice(range.end + syntax_end.length);
+        set(field, text);
+        field.setSelectionRange(range.start - syntax_start.length, range.end - syntax_start.length);
+        return;
+    } else if (is_inner_text_formatted(syntax_start, syntax_end, range, selected_text)) {
+        // Remove syntax inside the selection, if present.
+        text =
+            text.slice(0, range.start) +
+            linebreak_start +
+            text.slice(range.start + syntax_start.length, range.end - syntax_end.length) +
+            linebreak_end +
+            text.slice(range.end);
+        set(field, text);
+        field.setSelectionRange(range.start, range.end - syntax_start.length - syntax_end.length);
+        return;
+    }
+
+    // Otherwise, we don't have syntax within or around, so we add it.
+    wrapSelection(field, syntax_start, syntax_end);
+}
```

Comment:
This first refactoring commit is very hard to review, because it does like 5 things at once. 

I propose breaking it into 5 reviewable commits -- the end result can be the same, but it would be:
* First just move the code into new functions, passing parameters with the exact same variable names as we had before, so that `git show --color-moved` clearly shows no changes.
* Then a series of commits renaming variables, like `bold_syntax` => `syntax`.
* Then any functional changes.

Ideally with one such series for commit you're moving.

I like the names of the new functions, but I just can't tell what's changed.

This can be a new prep PR for just the broken out parts of this commit -- I think if we can integrate that, it'll be way easier to work through the rest.

Reply:
I broke up the 1st commit into 3:
- In the 1st I just move the bold formatting code into it's own function with almost no changes (except replacing `break` with `return` )
- the 2nd generalizes all variable / function names which were bold formatting specific earlier
- The 3rd changes the logic to make it more versatile for different types of formattings. Also improved the function interface a little (removed duplication of passing syntax when its same for start and end)

I could have broken up the 2nd one further, but since all the changes were of the same type, I let them be all together (and also coz the rebasing and splitting was getting a little difficult :grimacing:). Do let me know if it's alright now?

Reply:
The first two commits are great; the third is a bit hard to process, but felt at least possible to review -- going to merge it. 

I also kinda feel like it might be better to make these top-level functions -- i.e. passing parameters like `range` to them -- not sure but it seems like that might improve readability.

Reply:
I did consider making the formatting functions top level, but didn't due to multiple minor reasons:
- The pre-existing `format_list` function wasn't top level (so for consistency) 
- Passing in range, etc made it impossible for the `syntax_end` parameter to be optional (as only the last parameter can be optional)
- The function signature was getting a little too wordy I thought.

But these are minor issues, so if we decide to, moving these functions outside in the future should be easy too.